### DISCUSSION TITLE
Scorecard integration from Open Source Security Foundation

### DISCUSSION THREAD

**Discussion by jflowbird** [curl/curl#12609] [General]
Hi,

Me and a group of friends want to participate in the "Secure Open Source Rewards (SOS)" program from the Open Source Security Foundation. All of us are newbies in the open source comunity contribution.

Our idea is to start adding the "Scorecard" label in the README.md file and in the workflow. Then, try to improve the score working in the checks.

Is it a possible task that would be appreciated by the curl community?

Thanks in advance

**Comment by bagder** [comment]
> Our idea is to start adding the "Scorecard" label in the README.md file and in the workflow. Then, try to improve the score working in the checks.

Can you elaborate on what this means? What scorecard? What changes would that entail?

**Comment by cpswan** [comment]
#14319 introduced a section to `SECURITY.md` titled **OpenSSF Scorecard** which is actually about OpenSSF Best Practices (hence my #14635), but that does raise the question of whether curl should be creating an OpenSSF [Scorecard](https://scorecard.dev/) and the presence of this discussion shows that I'm not the first person to ask the question.

@bagder's comment in #14319 makes it clear that a badge wouldn't be added to show the score, but the underlying score data can be used in other ways, such as a report link (here's the one for [Scorecard itself](https://scorecard.dev/viewer/?uri=github.com/ossf/scorecard)) or [radar plot](https://chris.swanz.net/front-end-miniprojects/security-api/?repo=github.com/ossf/scorecard).

If I run the scorecard (v5.0.0) command line tool against the curl repo `scorecard --repo=https://github.com/curl/curl` then it comes back with a very respectable score of 8.3 (which would be a green badge if a badge was used). But some checks are obviously bad (because my GitHub token doesn't have access to non public repo data), so the score from running the [scorecard-action](https://github.com/ossf/scorecard-action) would likely be a good deal better.

The only change that would be entailed would (initially) adding the scorecard action workflow to the repo, and then adding a section about Scorecard to SECURITY.md.

Since a huge amount of effort has already been expended on Best Practices Gold, and underlying stuff like dependency pinning the remedial work to review and improve the score (which I expect to be in the mid 9s) is likely to be a bunch of 'won't fix' determinations. Though there's maybe scope for [SLSA](https://slsa.dev/) attestations and lifting the Signed-Releases score to 10. 

```text
Aggregate score: 8.3 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 3 / 10  | Branch-Protection      | branch protection is not       | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#branch-protection      |
|         |                        | maximal on development and all |                                                                                                                       |
|         |                        | release branches               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | CI-Tests               | no pull request found          | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#ci-tests               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | CII-Best-Practices     | badge detected: Gold           | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#cii-best-practices     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Code-Review            | Found 0/30 approved changesets | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#code-review            |
|         |                        | -- score normalized to 0       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Contributors           | project has 40 contributing    | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#contributors           |
|         |                        | companies or organizations     |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Dangerous-Workflow     | no dangerous workflow patterns | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#dangerous-workflow     |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Dependency-Update-Tool | update tool detected           | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Fuzzing                | project is fuzzed              | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 9 / 10  | License                | license file detected          | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Maintained             | 30 commit(s) and 19 issue      | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#maintained             |
|         |                        | activity found in the last 90  |                                                                                                                       |
|         |                        | days -- score normalized to 10 |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | packaging workflow not         | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#packaging              |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 7 / 10  | Pinned-Dependencies    | dependency not pinned by hash  | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#pinned-dependencies    |
|         |                        | detected -- score normalized   |                                                                                                                       |
|         |                        | to 7                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | SAST                   | SAST tool detected: CodeQL     | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#sast                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Security-Policy        | security policy file detected  | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#security-policy        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 8 / 10  | Signed-Releases        | 5 out of the last 5 releases   | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#signed-releases        |
|         |                        | have a total of 5 signed       |                                                                                                                       |
|         |                        | artifacts.                     |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 9 / 10  | Token-Permissions      | detected GitHub workflow       | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#token-permissions      |
|         |                        | tokens with excessive          |                                                                                                                       |
|         |                        | permissions                    |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | 0 existing vulnerabilities     | https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#vulnerabilities        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

**Reply by bagder** [reply]
And now you're going to get me yapping why the scorecard shows the wrong information for us and that's a reason I am against showing it off. I don't want to show something I don't agree with or understand.

- **branch protection is not maximal** - I find no documentation what this means and it sounds scary. I think we have the sufficient amount of protection. 3 out of 10 sounds terribly weak.
- **CI-tests** - not found? That sounds like a broken script. Look at any PR here, we run hundreds of CI jobs and many/most of them run thousands of tests each.
- **Code-Review**. 0 out of 10. That's just... wrong. I will admit that we do a whole lot of self-reviewing as maintainers, but again this sounds like wrong assumptions done by a script. I'm pretty sure a script cannot determine this accurately.
- **License**. 9 out of 10. We are fully REUSE compliant and we state license and copyright for each and every file present in git and tarballs. How can this *not* be a top score? What more can be done?
- **Signed-Releases** 8 out of 10. We have signed every single curl release for more than a decade. How can that be improved? How can you *sign more* than every release? Broken check?
- **SAST** - we use many more but you can't detect them with a script
- **Pinned-Dependencies** - what dependencies would this be? Why is this only a 7 if we have nothing to pin?

**Reply by cpswan** [reply]
@bagder I think the main problem here is that my GitHub token doesn't have sufficient permissions to correctly access information about the repo.

If you have a moment it would be really interesting to see the outcome of running the scorecard cli tool with your GitHub token. I'd expect the following:

* **branch protection** the tool will be able to get an accurate measure of branch protection settings. Though this might still not come out as a 10, as 'maximal' demands two approvers for each PR. I'd note that even scorecard itself doesn't score 10 on this measure.
* **CI-tests** and **Code-Review** obviously they're there, everybody can see them, but I think this is one of the areas where the check is using a permissioned part of the API that fails for a non maintainer.
* **License** I'd expect this to come out as a 10. Arguably it's too easy to get a 10 on this, because I see projects that haven't done per file licensing like curl that get full marks on this check.
* **Signed-Releases** this check expects to find a SLSA attestation for a 10, and only gives an 8 for other forms of signing. Is it worth opening another discussion thread on whether to implement SLSA for curl?
* **Pinned-Dependencies** the tool is scanning for all types of file that might define dependencies, and checking to see if they're pinned. Taking a casual poke around the repo for things I know it looks for: GitHub Actions Workflows - LGTM, Dockerfile - LGTM, Python [requirements.txt](tests/http/requirements.txt) (in tests/http) - no SHAs. There might be more, and if you implement the GitHub Action it populates the Security tab with recommendations to fix things. I'd also note that pinned-dependencies are a major source of toil. You already have Dependabot generating PRs for GitHub Actions dependencies, and that could be extended to Docker and Python etc. But it's going to generate a bunch of extra stuff for review. That can be ameliorated a little by adding groups to the dependabot.yml so at least you're not having to review and merge each bump individually.

I think with your token the score will come out in the high 9s, which is amazing.

I've never seen a 10, and arguably it's not very practical (the 2 reviewers for every PR thing). _Maybe_ curl could get to a 9.9 (which would be an industry leading score) by implementing SLSA attestations on releases.

_Disclaimer:_ apart from some trivial contributions I've had not part in defining or implementing Scorecard. I'm just a user of the tool who's found it a useful way to communicate that a project cares about security, and I've done a few talks showing other folk how to get started. With a Gold badge for OpenSSF Best practices curl is starting from an entirely different place versus most other projects, and most of the grind has been done already, which perhaps makes the residue all the more contentious.

**Reply by bagder** [reply]
There might be perfectly valid explanations why the scorecard is not better at figuring out the actual facts, but that is still a reason why I don't like to run it and show the number for curl as people will assume it is correct and an established truth. Why I believe the best practices badge is so much better.

On **dependency-pinning**: I'm not a fan of the seemingly indifference about the important difference between product and the development (CI). curl and libcurl do not have any dependencies to pin. The development environment (CI related stuff) have some. And btw, we already pin the dockerfile and much more.

I ran the scorecard cmdline tool locally but it gave the exact same 8.3 score as shown above. Maybe I'm holding it wrong.

> Is it worth opening another discussion thread on whether to implement SLSA for curl?

I don't know. To me it just feels like additional rubber-stamping with no real extra value. But I'm open to be educated and proven wrong.

**Reply by cpswan** [reply]
@badger did you run `scorecard --repo=https://github.com/curl/curl` with a `GITHUB_AUTH_TOKEN` in your environment that has (your maintainer) access to the repo? I'm really surprised that you also got 8.3. It implies some fundamental breakage that gives you every reason to question the reliability of the checks. You can add `--show-details` to see exactly what it's finding that it doesn't like, which is similar to the stuff that gets populated into the Security tab by the GitHub Action.

I just tried something different... running `scorecard --local=curl --show-details` against my local fork. That makes it skip a bunch of tests (because it can't reach into things like branch protection). But it spits out a score of 9.5, with the biggest whine being Pinned-Dependencies at 7/10, which I think is mostly Python stuff in CI 

I don't think there's _much_ value to pursuing SLSA for source bundles (apart from being a way to top up a Scorecard score). The whole thing is much more aimed at ensuring binaries and container images haven't been tampered with.

**Reply by bagder** [reply]
> with a GITHUB_AUTH_TOKEN in your environment that has (your maintainer) access to the repo?

I thought so and the tool did not complain about it. But...

> it spits out a score of 9.5

I suppose that is a pretty clear sign that something is insufficient with my token. So I *was* holding it wrong!

The 9/10 for **License** is said to be because *project license file does not contain an FSF or OSI license.* - which I guess might be technically correct, but I think misleading. The curl license is in effect MIT.

The 7/10 for **Pinned-Dependencies** is exactly as said before: a bunch of CI jobs. All of them are fire-and-forget jobs with nothing getting back to the project, so worst case is ruined tests. The scorecard check seems to not understand this concept which I find weak.

> The whole thing is much more aimed at ensuring binaries and container images haven't been tampered with

Shouldn't it be an even higher score when you don't even ship binaries or container images so therefore not needing those things? Instead we get a lower score for not having them. Again this feeling that this scorecard thing is designed for another kind of project.

All together: the items the scorecard lists could perhaps be worthy of mentioning and listing in documentation, but I don't think the scorecard output itself brings much good to include.

Related: recently I added a new section to the everything curl book about how we work to mitigate [Backdoors and supply chain risks](https://everything.curl.dev/project/security.html#backdoors-and-supply-chain-risks)

**Reply by cpswan** [reply]
A couple more thoughts come to mind:

### 1. An explicit statement about objections to the Scorecard measures

A section on Scorecard in the `SECURITY.md` would provide a place to explain some of the positions that are coming out in this thread, about why curl might choose to diverge from the things that Scorecard looks for.

For better or worse people will run the tool against the project and draw their own conclusions. If they're diligent they might find this discussion. But a statement in `SECURITY.md` would likely be more visible. For example a bunch of security researchers used Scorecard in their paper [Exploring the Threat of Software Supply Chain Attacks on Containerized Applications](https://www.researchgate.net/publication/374838986_Exploring_the_Threat_of_Software_Supply_Chain_Attacks_on_Containerized_Applications#fullTextFileContent). curl/libcurl didn't feature in that analysis (as they were looking at popular projects with very low scores), but another time, with another bunch of people, they might just grab a top X list and curl/libcurl is sure to feature.

Also Scorecard is itself open source. flawed, and malleable to fix what's broken, so I'm sure that things like the license checker could be fixed up. I'd also expect that the Scorecard team would be keen to engage with a flagship project like curl, particularly given the effort that's already been expended on Best Practices.

### 2. Use the tool as a security linter

Even if you choose not to publish a score, there could be some utility in having the Scorecard tool as part of CI.

If we stick with **Pinned-Dependencies** for a moment then we can look back and see that #13628 introduced dependency pinning for GitHub Actions. I've never used Renovate, but my guess would be that this was some manual intervention through that tool.

@cmeister2 [justifies](https://github.com/curl/curl/pull/13628#issuecomment-2111231781) the change:

> My understanding is that pinning dependencies is preferred for OpenSSF, which I believe curl is signed up for. It's also better to pin to a hash rather than a tag because tags can be changed on Git.

@bagder is right to be [skeptical](https://github.com/curl/curl/pull/13628#issuecomment-2111654520):

> How often is this changing, do we know? Does it risk becoming a nuisance?

In my experience pinning Actions causes a fair amount of toil. The codeql-action is one of the worst culprits, with bumps every few days.

But it's decided by @bagder to [proceed](https://github.com/curl/curl/pull/13628#pullrequestreview-2057730042), at least for now:

> Let's do this. If it causes too much churn or nuisance we can reconsider later.

And then a few weeks later https://github.com/curl/curl/commit/f99c08dba40307c07341013ff5f71fa8e3464ffc comes along and introduces an unpinned Action:

```yaml
      - name: 'vcpkg cache setup'
        uses: actions/github-script@v7
```

Renovate seemingly isn't configured to jump on this and get it in line with the others, and so the decision made in mid May to pin Actions is already crumbling by early June. This is something that Scorecard could help to catch.

Checklists like Best Practices are great at capturing a point in time snapshot of where things were. But as noted elsewhere it takes regularly running scripts to stay on top of things. e.g. 'The project MUST have performed a security review within the last 5 years.' was true when the questionnaire was last updated in 2021-07-02: 

![image](https://github.com/user-attachments/assets/7a355cf8-c912-4e9d-b6ab-0bcec429556f)

and an update to that might point to the [curl security audit reports](https://curl.se/docs/audits.html) that links to more recent reviews.

Scorecard doesn't actually help with that particular thing, which is why Scorecard still includes Best Practices, but there's a bunch of other stuff where it's not simply taking a point in time assertion that something is happening, it's checking that those things are continuing to happen. Another bit of verify in trust but verify.

**Reply by icing** [reply]
There seems to be a tension between "make the tool work so the project score shows correctly" and "make the project jump through hoops so the tool shows what we want to see". I am not a fan of the latter approach. It does not scale. I do not consider it a worthwhile endeavour.

**Reply by bagder** [reply]
> For better or worse people will run the tool against the project and draw their own conclusions

Sure. People run every imaginable tool against curl and draw conclusions based on that. I think our emphasis should be to provide clear and accessible documentation that explains the situation in a tool independent fashion. There are just too many tools.

> Also Scorecard is itself open source

I realize that I should file issues and "complain" to the scorecard team rather than ventilating here if we want the issues fixed. I simply have not had time and energy to do this.

> as a security linter

That's not a bad idea, but the way the tool works it seems complicated? I mean, we would like a test to be green when we add it and then go red if we break something. Can it do that?

Pinning again: please explain the security problem our lack of pinning in CI jobs cause. The risk that someone can plant an infected software versions used in tests that deliberately make specific test cases go green when they in reality had faults and this way someone can trick us to merge a faulty PR? Something that would be immediately detected when the tests run locally and in the other CI services.

Pinning was added to the CI jobs partly or mostly to "silence" tools that check for this, but since we're not doing it 100% tools will still complain so right now it feels like we get the worst of both worlds. I now mostly question why we do pinning at all.

> The project MUST have performed a security review within the last 5 years

Seems like a feature-request to the best practices project then to alert people? BTW: it is still valid since we have had repeated audits even if we have not resubmitted that form.

**Reply by cpswan** [reply]
Taking a slightly different tack...

What was persuasive about the Core Infrastructure Initiative (as was) Best Practices that got you on board? Was there straightforward philosophical alignment, or did it take a bit of compromise (maybe on both sides)? As I browse through the responses there's only one where curl noped out, and it turned out not to be a blocker:

![image](https://github.com/user-attachments/assets/faa2ac90-9aaa-444f-8bea-5d6174e5a265)

I think there's stuff in motion for Scorecard users to mark checks as N/A for their project, and that would be particularly relevant for something like curl that doesn't ship binaries (as a bunch of the checks are there to counter tampering in the binary pipeline).

***

Returning to your questions:

> Can it do that?

Maybe. But it would be bending things out of shape, which probably isn't a good idea.

> trick us to merge a faulty PR?

The complexity of a supply chain attack to do this to curl would be pretty mind boggling. Maybe (at least) an order of magnitude more so that xzutils (and that was a painstaking and inordinately complex attack). On face value I'd be much more worried about the (as yet unpinned) Python dependencies, but I don't know the project well enough to judge how important they are. Pinning also implies some degree of careful evaluation of each bump, but reviewers are poorly equipped to do that. There's also a herd immunity angle here - curl doesn't have to do this so long as the surrounding ecosystem is paying attention (and post xzutils there's a LOT more focus on this stuff).

> Seems like a feature-request to the best practices project then to alert people?

Somebody from OpenSSF might jump in to correct me here, but I think the reality is that Best Practices will get subsumed by Scorecard over time (because ongoing checks are better than one off questionnaires). There will of course be a residue of things that can't be checked in an easy/standard way (and "The project MUST have performed a security review within the last 5 years" falls into that category), so over time I'd expect a diminishing questionnaire. It would make sense then to facilitate a review cycle, particularly for stuff that's time bounded. I suspect that legislation like the Cyber Resilience Act will also come into play here. There's been some sterling work done to ensure that the legislation don't put undue pressure onto open source projects. But things like '[So the Department of Energy emailed me](https://daniel.haxx.se/blog/2024/08/14/so-the-department-of-energy-emailed-me/)' will keep happening, and it might be easier for all concerned if questions are answered once in a place that everybody can refer to.

**Reply by bagder** [reply]
> What was persuasive about the Core Infrastructure Initiative (as was) Best Practices that got you on board

It came at the right time and presented a good checklist that when we had them met, we could be fairly sure that we actually follow most open source best practices. It pushed us/me into fixing several things/processes we previously had not gotten around to do or document. I'll admit that the last bunch of the requirements became a little bit of a challenge to me to see if we could actually take it all the way to the top.

> things like '[So the Department of Energy emailed me](https://daniel.haxx.se/blog/2024/08/14/so-the-department-of-energy-emailed-me/)' will keep happening

Those things will unfortunately continue to happen no matter what we do I believe.

**Comment by bagder** [comment]
The scorecard project has over 300 open issues and does not give an impression of being very alive. I provided feedback upstream on two of the issues:

- [Exaggerating the importance of pinning in CIs](https://github.com/ossf/scorecard/discussions/4311)
- [update the SPDX license list](https://github.com/ossf/scorecard/issues/4031#issuecomment-2309607571)
